### ISSUE TITLE
Crash/assert on collision between connection reattempt and returning error

### ISSUE THREAD

**Issue by shinrich** [apache/trafficserver#3265]
We have seen this crash on our earlier 7.1.1-based build, but looking at the code in the area, I think this is likely an issue for 7.1.2 as well. 

Looking at history and the milestones, it appears that the SM machine made an attempt to connect to origin.  It connects but the origin sends an EOS.  The state machine sets up a tunnel to return an error message, but it incorrectly continues on with connection reattempts.

I think the problem is in HttpSM::state_read_servrer_resposne_header in the VC_EVENT_EOS csae.  We carefully detect that 0 bytes have been written and set up an error response and return so we don't accidentally process this as a HTTP 0.9 response.  I assume this path doesn't appropriately set enough state to stop the connection retries.  Since we no longer support 0.9, we don't need this check and can let the logic pass through.  Later, we set the state to PARSE_RESULT_ERROR if we parse a 0.9 response on the not first response on a connection.  We could change that to return PARSE_RESULT_ERROR for all responses on a connection.  Then it will go through the normal error case which should stop any connection retries.

```
[ 0  ] libc-2.12.so       raise                                      ( :undefined                ) 
[ 1  ] libc-2.12.so       __GI_abort                                 ( :undefined                ) 
[ 2  ] libtsutil.so.7     ink_abort                                  ( ink_error.cc:99           ) 
[ 3  ] libtsutil.so.7     _ink_assert                                ( ink_assert.cc:37          ) 
[ 4  ] traffic_server     HttpSM::state_http_server_open(int, void*) ( HttpSM.cc:1928            ) 
[ 5  ] traffic_server     HttpSM::main_handler(int, void*)           ( HttpSM.cc:2797            ) 
[ 6  ] traffic_server     Continuation::handleEvent(int, void*)      ( I_Continuation.h:153      ) 
[ 7  ] traffic_server     HttpTunnel::main_handler(int, void*)       ( HttpTunnel.cc:1648        ) 
[ 8  ] traffic_server     Continuation::handleEvent(int, void*)      ( I_Continuation.h:153      ) 
[ 9  ] traffic_server     write_signal_and_update                    ( UnixNetVConnection.cc:175 ) 
[ 10 ] traffic_server     write_signal_done                          ( UnixNetVConnection.cc:217 ) 
[ 11 ] traffic_server     write_to_net_io                            ( UnixNetVConnection.cc:564 ) 
[ 12 ] traffic_server     write_to_net                               ( UnixNetVConnection.cc:416 ) 
[ 13 ] traffic_server     NetHandler::waitForActivity(long)          ( UnixNet.cc:514            ) 
[ 14 ] traffic_server     EThread::execute_regular()                 ( UnixEThread.cc:232        ) 
[ 15 ] traffic_server     EThread::execute()                         ( UnixEThread.cc:262        ) 
[ 16 ] traffic_server     spawn_thread_internal                      ( Thread.cc:85              ) 
[ 17 ] libpthread-2.12.so start_thread                               ( :undefined                )
```

**Comment by shinrich** [comment]
The following stack also looks similar.

```
[ 0  ] libc-2.12.so       raise                                                   ( :undefined           ) 
[ 1  ] libc-2.12.so       __GI_abort                                              ( :undefined           ) 
[ 2  ] libtsutil.so.7     ink_abort                                               ( ink_error.cc:99      ) 
[ 3  ] libtsutil.so.7     _ink_assert                                             ( ink_assert.cc:37     ) 
[ 4  ] traffic_server     HttpSM::state_http_server_open(int, void*)              ( HttpSM.cc:1928       ) 
[ 5  ] traffic_server     HttpSM::main_handler(int, void*)                        ( HttpSM.cc:2797       ) 
[ 6  ] traffic_server     Continuation::handleEvent(int, void*)                   ( I_Continuation.h:153 ) 
[ 7  ] traffic_server     HttpTunnel::main_handler(int, void*)                    ( HttpTunnel.cc:1648   ) 
[ 8  ] traffic_server     Continuation::handleEvent(int, void*)                   ( I_Continuation.h:153 ) 
[ 9  ] traffic_server     Http2Stream::main_event_handler(int, void*)             ( Http2Stream.cc:90    ) 
[ 10 ] traffic_server     Continuation::handleEvent(int, void*)                   ( I_Continuation.h:153 ) 
[ 11 ] traffic_server     EThread::process_event(Event*, int)                     ( UnixEThread.cc:122   ) 
[ 12 ] traffic_server     EThread::process_queue(Queue<Event, Event::Link_link>*) ( UnixEThread.cc:159   ) 
[ 13 ] traffic_server     EThread::execute_regular()                              ( UnixEThread.cc:191   ) 
[ 14 ] traffic_server     EThread::execute()                                      ( UnixEThread.cc:262   ) 
[ 15 ] traffic_server     spawn_thread_internal                                   ( Thread.cc:85         ) 
[ 16 ] libpthread-2.12.so start_thread                                            ( :undefined           )
```

**Comment by shinrich** [comment]
Addressed by PR #3285
